File: ts-sdk/src/exporter.ts
import { AgentScopeClient } from "./client";
import type { IngestPayload, TelemetryExporterLike } from "./types";

const pendingExports = new Set<Promise<void>>();

export class TelemetryExporter implements TelemetryExporterLike {
  constructor(private readonly client: AgentScopeClient = new AgentScopeClient()) {}

  export(payload: IngestPayload): Promise<void> {
    const task = this.client.ingest(payload).finally(() => {
      pendingExports.delete(task);
    });
    pendingExports.add(task);
    return task;
  }
}

export async function flushPendingExports(): Promise<void> {
  while (pendingExports.size > 0) {
    await Promise.all([...pendingExports]);
  }
}

File: python-sdk/agentscope/exporter.py
from typing import Any, Dict, List

from .client import AgentScopeClient


class TelemetryExporter:
    def __init__(self, client: AgentScopeClient | None = None) -> None:
        self.client = client or AgentScopeClient()

    def export(self, run: Dict[str, Any], spans: List[Dict[str, Any]], artifacts: List[Dict[str, Any]]) -> None:
        payload = {
            "run": run,
            "spans": spans,
            "artifacts": artifacts,
        }
        self.client.ingest(payload)
